clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -783,8 +783,10 @@ public void handle(HeartbeatResponse heartbeatResponse, RequestFuture<Void> futu
         @Override
         public void onFailure(RuntimeException e, RequestFuture<T> future) {
             // mark the coordinator as dead
-            if (e instanceof DisconnectException)
+            if (e instanceof DisconnectException) {
                 coordinatorDead();
+                requestRejoin();
```

Comment:
I'm not sure if we should set `requestRejoin` in the base class (`CoordinatorResponseHandler `). For example, `HeartbeatResponseHandler` also extends from it, but for that request if we get a disconnect, we should just mark the coordinator as dead in order to re-discover it; and then after new coordinator rediscovered retry sending heartbeat request and if that succeed just proceed as normal. Setting it here will force heartbeat request disconnection to also trigger a join group.